packages/orm/src/Model.ts

```diff
@@ -0,0 +1,786 @@
+
+        if (field instanceof ForeignKey || field instanceof OneToOne) {
+          // update one-one/fk relations
+          mergeObj[mergeKey] = normalizeEntity(mergeObj[mergeKey]);
+        } else if (field instanceof ManyToMany) {
+          // field is forward relation
+          m2mRelations[mergeKey] = mergeObj[mergeKey];
+          delete mergeObj[mergeKey];
+        }
+      } else if (virtualFields.hasOwnProperty(mergeKey)) {
+        const field = virtualFields[mergeKey];
+        if (field instanceof ManyToMany) {
+          // field is backward relation
+          m2mRelations[mergeKey] = mergeObj[mergeKey];
+          delete mergeObj[mergeKey];
+        }
+      }
+    }
+
+    const mergedFields = {
+      ...this._fields,
+      ...mergeObj,
+    };
+
+    const updatedModel = new ThisModel(this._fields);
+    updatedModel._initFields(mergedFields); // eslint-disable-line no-underscore-dangle
+
+    // determine if model would have different related models after update
+    updatedModel._refreshMany2Many(m2mRelations); // eslint-disable-line no-underscore-dangle
+    const relationsEqual = Object.keys(m2mRelations).every(
+      (name) =>
+        !arrayDiffActions(
+          (this as ModelData)[name],
+          (updatedModel as ModelData)[name]
+        )
+    );
+    const fieldsEqual = this.equals(updatedModel);
+
+    // only update fields if they have changed (referentially)
+    if (!fieldsEqual) {
+      this._initFields(mergedFields);
+    }
+
+    // only update many-to-many relationships if any reference has changed
+    if (!relationsEqual) {
+      this._refreshMany2Many(m2mRelations);
+    }
+
+    // only apply the update if a field or relationship has changed
+    if (!fieldsEqual || !relationsEqual) {
+      ThisModel.session.applyUpdate({
+        action: UPDATE,
+        query: getByIdQuery(this),
+        payload: mergeObj,
+      });
+    }
+  }
+
+  /**
+   * Updates {@link Model} instance attributes to reflect the
+   * database state in the current session.
+   * @return {undefined}
+   */
+  refreshFromState() {
+    this._initFields(this.ref);
+  }
+
+  /**
+   * Deletes the record for this {@link Model} instance.
+   * You'll still be able to access fields and values on the instance.
+   *
+   * @return {undefined}
+   */
+  delete() {
+    const ThisModel = this.getClass();
+    if (typeof ThisModel.session === "undefined") {
+      throw new Error(
+        [
+          `Tried to delete a ${ThisModel.modelName} model instance without a session. `,
+          "You cannot call `.delete` on an instance that you did not receive from the database.",
+        ].join("")
+      );
+    }
+
+    this._onDelete();
+    ThisModel.session.applyUpdate({
+      action: DELETE,
+      query: getByIdQuery(this),
+    });
+  }
+
+  /**
+   * Update many-many relations for model.
+   * @param relations
+   * @return undefined
+   * @private
+   */
+  _refreshMany2Many(relations: Record<string, object>) {
```

Comment:
as mentioned above - maybe we could use the TS `private` keyword to actually enforce it (and perhaps remove the `_` prefix)?